### PR TITLE
Modify StorageImplCreateHelper

### PR DESCRIPTION
**Pull request by CLiqing**
I want to use tensor.untyped_storage()[a:b] for ``PrivateUse1`` backend but fail. The code will go into ``THPStorage_get``:
https://github.com/pytorch/pytorch/blob/bb6eba189f92790f278f1ee2840f3786da02acaa/torch/csrc/Storage.cpp#L525-L540

Here ``torch`` will create a new ``c10::StorageImpl`` but not consider about ``PrivateUse1`` backend.

### CODE REVIEW

**Comment by ezyang** [comment]
Is it possible to write a test that exercises the fixed behavior?

NGL, the code change here looks plausible but I am not entirely sure what it is doing.

**Comment by CLiqing** [comment]
> Is it possible to write a test that exercises the fixed behavior?
> 
> NGL, the code change here looks plausible but I am not entirely sure what it is doing.

I think there is no need to add additional test, because we found this problem when testing test_torch.py.
https://github.com/pytorch/pytorch/blob/da0635d17c8fc777010fc3a2c5efedfade499432/test/test_torch.py#L192-L194
This code will check tensor.untyped_storage()[a:b].

For the second problem, both ``THPStorage_get`` and ``THPStorage_pynew`` need to make a new ``StorageImpl``, but ``THPStorage_get`` does not consider about other devices while ``THPStorage_pynew`` does. We want to reuse ``make_storage_impl`` used in ``THPStorage_new`` into ``THPStorage_get``. However, ``make_storage_impl`` does not support ``StorageImpl's constructor with at::DataPtr``, so we did such code change.

**Comment by ezyang** [comment]
Well, it would still be profitable to replicate the test_torch.py test, but using PrivateUse1, no?

**Comment by CLiqing** [comment]
After re-reading https://github.com/pytorch/pytorch/pull/100237 and incorporating https://github.com/pytorch/pytorch/issues/119146 's requirements, I submitted a new version. Previous ``make_storage_impl`` is mainly moved from ``THPStorage_pynew``. For generality, I put ``allocator_opt`` back to ``THPStorage_pynew``, and only save code of creating ``c10::StorageImpl``. In addition, I move ``make_storage_impl`` to ``StorageImpl.h`` so it is together with the ``PrivateUse1`` registration and can be used in other APIs .
I think this method is more reasonable, compatible with previous functions and new ``PrivateUse1`` requirements. What do you think @ezyang ?

**Comment by CLiqing** [comment]
@ezyang , I have fixed some CI problems, can you please run again?

**Comment by CLiqing** [comment]
@albanD , could u please take a look at this pr? Thanks!

**Comment by CLiqing** [comment]
@ezyang , what do u think of the new implemention?

**Comment by CLiqing** [comment]
@ezyang , could u please take a look at this pr?

File: test/cpp_extensions/open_registration_extension.cpp [hunk 1]
```diff
@@ -137,10 +137,17 @@ void custom_set_backend_meta(const at::Tensor& t) {
...
 c10::intrusive_ptr<c10::StorageImpl> make_custom_storage_impl(c10::StorageImpl::use_byte_size_t,
                                                               c10::SymInt size_bytes,
+                                                              c10::DataPtr data_ptr,
                                                               c10::Allocator* allocator,
                                                               bool resizable) {
-  c10::intrusive_ptr<c10::StorageImpl> custom_storage_impl = c10::make_intrusive<c10::StorageImpl>(
+  c10::intrusive_ptr<c10::StorageImpl> custom_storage_impl;
+  if (data_ptr == nullptr){
+    custom_storage_impl = c10::make_intrusive<c10::StorageImpl>(
       c10::StorageImpl::use_byte_size_t(), size_bytes, allocator, resizable);
```
Comment (ezyang) [inline test/cpp_extensions/open_registration_extension.cpp:146]: This is kind of perplexing, why not just pass in the nullptr and have the constructor deal with it

**Comment by CLiqing** [comment]
I have tranfered data_ptr==nullptr condition to the constructor. @ezyang , is there any other suggestions?

File: test/cpp_extensions/open_registration_extension.cpp (hunk 1 above)
Comment (CLiqing) [reply test/cpp_extensions/open_registration_extension.cpp]: > This is kind of perplexing, why not just pass in the nullptr and have the constructor deal with it

I thought constructor of StorageImpl is a basic function and don't want to modify it at first. Now I allowed  ``data_ptr`` to be nullptr in StorageImpl constructor, could you please take some time to look at the new commits, @ezyang ?

File: torch/csrc/Storage.cpp [hunk 2]
```diff
@@ -290,29 +290,50 @@ static void THPStorage_subclass_dealloc(PyObject* self) {
...
 }
 
-c10::intrusive_ptr<c10::StorageImpl> make_storage_impl(
-    c10::StorageImpl::use_byte_size_t use_byte_size,
-    c10::SymInt size_bytes,
-    c10::Allocator* allocator,
-    bool resizable,
-    c10::optional<int64_t> allocator_opt,
-    c10::optional<at::Device> device_opt) {
+static PyObject* THPStorage_pynew(
```
Comment (ezyang) [inline torch/csrc/Storage.cpp:293]: Was moving this function really necessary? It makes it hard to read the diff

Comment (CLiqing) [reply torch/csrc/Storage.cpp:293]: > Was moving this function really necessary? It makes it hard to read the diff

I think it is necessary. These changes are mainly from reverting https://github.com/pytorch/pytorch/pull/100237. That pr extends storage creatation for custom storageImpl, but it doesn't consider other codes may need it like ``new_qtensor``. In my opinion, moving ``make_storage_impl`` to c10/core/StorageImpl.h is better for generality, @ezyang.

**Comment by ezyang** [comment]
@albanD approved the original PR that added the storage creation hook, so I am deferring this review to him.

**Comment by CLiqing** [comment]
@albanD , could u please take some time and have a look at this pr? Thanks!

File: torch/csrc/Storage.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/Storage.cpp:293]: I would agree with ezyang that the diff is pretty hard to read here.
You seems to be moving part of make_storage_impl into THPStorage_pynew below and the other half in c10?
It does change the order of quite a bit of the logic in here...

File: c10/core/StorageImpl.cpp [hunk 3]
```diff
@@ -36,4 +36,39 @@ StorageImplCreateHelper GetStorageImplCreate(DeviceType t) {
   return StorageImplCreate[device_type];
 }
 
+c10::intrusive_ptr<c10::StorageImpl> make_storage_impl(
+    c10::StorageImpl::use_byte_size_t use_byte_size,
+    c10::SymInt size_bytes,
+    c10::DataPtr data_ptr,
+    c10::Allocator* allocator,
+    bool resizable,
+    c10::optional<at::Device> device_opt) {
```
Comment (albanD) [inline c10/core/StorageImpl.cpp:45]: Do I read correctly that both the DataPtr argument and this optional Device were added here?
Given that DataPtr has a device attribute already and a few of the call site don't have it, why not just have an optional DataPtr?

Worst case, the allocator can definitely tell you which device it allocates for

File: c10/core/StorageImpl.h [hunk 4]
```diff
@@ -55,6 +55,11 @@ struct C10_API StorageImpl : public c10::intrusive_ptr_target {
         resizable_(resizable),
         received_cuda_(false),
         allocator_(allocator) {
+    if (data_ptr_ == nullptr) {
```
Comment (albanD) [inline c10/core/StorageImpl.h:58]: I don't think we want to allow empty dataptr and allocate under the hood here. That is a significant semantic creep for this function

Comment (CLiqing) [reply c10/core/StorageImpl.h:58]: > I don't think we want to allow empty dataptr and allocate under the hood here. That is a significant semantic creep for this function

Ezyang suggested “pass in the nullptr and have the constructor deal with it” so I have changed to this. Maybe rollback such code is better?

File: c10/core/StorageImpl.cpp (hunk 3 above)
Comment (CLiqing) [reply c10/core/StorageImpl.cpp:45]: > Do I read correctly that both the DataPtr argument and this optional Device were added here? Given that DataPtr has a device attribute already and a few of the call site don't have it, why not just have an optional DataPtr?
> 
> Worst case, the allocator can definitely tell you which device it allocates for

Same to ezyang's suggestion, here DataPtr is allowed to be nullptr so will not be an optional variable.

**Comment by CLiqing** [comment]
@albanD , would u plz take a look at new changes?

**Comment by CLiqing** [comment]
@albanD , would u plz take a look at new changes?

**Review by albanD — APPROVED** [review]
Sorry about that, this change is pretty dangerous and hard to parse so taking quite a bit of time to review.
That sounds good to me now, thanks!
